File: core-python/src/rappel_core/dag/builder/conditionals.py
"""Conditional conversion helpers."""

from __future__ import annotations

import copy
from typing import List, Optional, Sequence

from proto import ast_pb2 as ir

from ..models import ConvertedSubgraph, DagConversionError, DAGEdge
from ..nodes import BranchNode, JoinNode


class ConditionalConversionMixin:
    """Convert conditional blocks into branch/join nodes."""

    def convert_conditional(self, cond: ir.Conditional) -> ConvertedSubgraph:
        """Convert an if/elif/else tree into a branch + optional join graph."""
        nodes: List[str] = []

        branch_id = self.next_id("branch")
        branch_node = BranchNode(
            id=branch_id,
            description="branch",
            function_name=self.current_function,
        )
        self.dag.add_node(branch_node)
        nodes.append(branch_id)

        if not cond.HasField("if_branch"):
            raise DagConversionError("conditional missing if_branch")
        if_branch = cond.if_branch
        if not if_branch.HasField("condition"):
            raise DagConversionError("if branch missing guard expression")
        if_guard = copy.deepcopy(if_branch.condition)

        if_body_graph = (
            self.convert_block(if_branch.block_body)
            if if_branch.HasField("block_body")
            else ConvertedSubgraph.noop()
        )
        nodes.extend(if_body_graph.nodes)

        prior_guards = [if_guard]
        elif_graphs: List[tuple[ir.Expr, ConvertedSubgraph]] = []

        for elif_branch in cond.elif_branches:
            if not elif_branch.HasField("condition"):
                raise DagConversionError("elif branch missing guard expression")
            elif_cond = copy.deepcopy(elif_branch.condition)
            compound_guard = self.build_compound_guard(prior_guards, elif_cond)
            prior_guards.append(elif_cond)

            graph = (
                self.convert_block(elif_branch.block_body)
                if elif_branch.HasField("block_body")
                else ConvertedSubgraph.noop()
            )
            nodes.extend(graph.nodes)
            elif_graphs.append((compound_guard, graph))

        else_graph = (
            self.convert_block(cond.else_branch.block_body)
            if cond.HasField("else_branch") and cond.else_branch.HasField("block_body")
            else ConvertedSubgraph.noop()
        )
        nodes.extend(else_graph.nodes)

        join_needed = (
            if_body_graph.is_noop
            or bool(if_body_graph.exits)
            or any(graph.is_noop or graph.exits for _, graph in elif_graphs)
            or else_graph.is_noop
            or bool(else_graph.exits)
        )

        join_id: Optional[str] = None
        if join_needed:
            join_id = self.next_id("join")
            join_node = JoinNode(
                id=join_id,
                description="join",
                function_name=self.current_function,
            )
            self.dag.add_node(join_node)
            nodes.append(join_id)

        def connect_guarded_branch(
            guard: ir.Expr, graph: ConvertedSubgraph, join_target: Optional[str]
        ) -> None:
            if graph.is_noop:
                if join_target:
                    self.dag.add_edge(
                        DAGEdge.state_machine_with_guard(branch_id, join_target, guard)
                    )
                return

            if graph.entry:
                self.dag.add_edge(DAGEdge.state_machine_with_guard(branch_id, graph.entry, guard))

            if join_target:
                for exit_node in graph.exits:
                    self.dag.add_edge(DAGEdge.state_machine(exit_node, join_target))

        def connect_else_branch(graph: ConvertedSubgraph, join_target: Optional[str]) -> None:
            if graph.is_noop:
                if join_target:
                    self.dag.add_edge(DAGEdge.state_machine_else(branch_id, join_target))
                return

            if graph.entry:
                self.dag.add_edge(DAGEdge.state_machine_else(branch_id, graph.entry))

            if join_target:
                for exit_node in graph.exits:
                    self.dag.add_edge(DAGEdge.state_machine(exit_node, join_target))

        connect_guarded_branch(if_guard, if_body_graph, join_id)
        for guard, graph in elif_graphs:
            connect_guarded_branch(guard, graph, join_id)
        connect_else_branch(else_graph, join_id)

        return ConvertedSubgraph(
            entry=branch_id,
            exits=[join_id] if join_id else [],
            nodes=nodes,
            is_noop=False,
        )

    def build_compound_guard(
        self, prior_guards: Sequence[ir.Expr], current_condition: Optional[ir.Expr]
    ) -> ir.Expr:
        """Build a guard for elif branches: not prior_guards and current_condition."""
        parts: List[ir.Expr] = []
        for guard in prior_guards:
            parts.append(
                ir.Expr(
                    unary_op=ir.UnaryOp(
                        op=ir.UnaryOperator.UNARY_OP_NOT,
                        operand=copy.deepcopy(guard),
                    )
                )
            )

        if current_condition is not None:
            parts.append(copy.deepcopy(current_condition))

        if not parts:
            raise DagConversionError(
                "build_compound_guard called with no prior conditions and no current condition"
            )
        if len(parts) == 1:
            return parts[0]

        result = parts.pop(0)
        for part in parts:
            result = ir.Expr(
                binary_op=ir.BinaryOp(
                    left=copy.deepcopy(result),
                    op=ir.BinaryOperator.BINARY_OP_AND,
                    right=copy.deepcopy(part),
                )
            )
        return result

File: crates/lib/dag/src/builder/conditionals.rs
//! Conditional conversion helpers.

use waymark_proto::ast as ir;

use super::super::models::{ConvertedSubgraph, DAGEdge, DagConversionError};
use super::super::nodes::{BranchNode, JoinNode};
use super::converter::DAGConverter;

/// Convert conditional blocks into branch/join nodes.
impl DAGConverter {
    /// Convert an if/elif/else tree into a branch + optional join graph.
    pub fn convert_conditional(
        &mut self,
        cond: &ir::Conditional,
    ) -> Result<ConvertedSubgraph, DagConversionError> {
        let mut nodes: Vec<String> = Vec::new();

        let branch_id = self.next_id("branch");
        let branch_node =
            BranchNode::new(branch_id.clone(), "branch", self.current_function.clone());
        self.dag.add_node(branch_node.into());
        nodes.push(branch_id.clone());

        let if_branch = cond
            .if_branch
            .as_ref()
            .ok_or_else(|| DagConversionError("conditional missing if_branch".to_string()))?;
        let if_guard = if_branch
            .condition
            .as_ref()
            .ok_or_else(|| DagConversionError("if branch missing guard expression".to_string()))?
            .clone();

        let if_body_graph = if let Some(block) = &if_branch.block_body {
            self.convert_block(block)?
        } else {
            ConvertedSubgraph::noop()
        };
        nodes.extend(if_body_graph.nodes.clone());

        let mut prior_guards = vec![if_guard.clone()];
        let mut elif_graphs: Vec<(ir::Expr, ConvertedSubgraph)> = Vec::new();

        for elif_branch in &cond.elif_branches {
            let elif_cond = elif_branch
                .condition
                .as_ref()
                .ok_or_else(|| {
                    DagConversionError("elif branch missing guard expression".to_string())
                })?
                .clone();
            let compound_guard = self.build_compound_guard(&prior_guards, Some(&elif_cond))?;
            prior_guards.push(elif_cond);

            let graph = if let Some(block) = &elif_branch.block_body {
                self.convert_block(block)?
            } else {
                ConvertedSubgraph::noop()
            };
            nodes.extend(graph.nodes.clone());
            elif_graphs.push((compound_guard, graph));
        }

        let else_graph = if let Some(else_branch) = &cond.else_branch {
            if let Some(block) = &else_branch.block_body {
                self.convert_block(block)?
            } else {
                ConvertedSubgraph::noop()
            }
        } else {
            ConvertedSubgraph::noop()
        };
        nodes.extend(else_graph.nodes.clone());

        let join_needed = if_body_graph.is_noop
            || !if_body_graph.exits.is_empty()
            || elif_graphs
                .iter()
                .any(|(_, graph)| graph.is_noop || !graph.exits.is_empty())
            || else_graph.is_noop
            || !else_graph.exits.is_empty();

        let mut join_id: Option<String> = None;
        if join_needed {
            let join_node_id = self.next_id("join");
            let join_node = JoinNode::new(
                join_node_id.clone(),
                "join",
                None,
                None,
                self.current_function.clone(),
            );
            self.dag.add_node(join_node.into());
            nodes.push(join_node_id.clone());
            join_id = Some(join_node_id);
        }

        let connect_guarded_branch =
            |converter: &mut DAGConverter,
             guard: ir::Expr,
             graph: ConvertedSubgraph,
             join_target: Option<&String>| {
                if graph.is_noop {
                    if let Some(join) = join_target {
                        converter
                            .dag
                            .add_edge(DAGEdge::state_machine_with_guard(&branch_id, join, guard));
                    }
                    return;
                }
                if let Some(entry) = graph.entry.as_ref() {
                    converter
                        .dag
                        .add_edge(DAGEdge::state_machine_with_guard(&branch_id, entry, guard));
                }
                if let Some(join) = join_target {
                    for exit_node in &graph.exits {
                        converter
                            .dag
                            .add_edge(DAGEdge::state_machine(exit_node, join));
                    }
                }
            };

        let connect_else_branch = |converter: &mut DAGConverter,
                                   graph: ConvertedSubgraph,
                                   join_target: Option<&String>| {
            if graph.is_noop {
                if let Some(join) = join_target {
                    converter
                        .dag
                        .add_edge(DAGEdge::state_machine_else(&branch_id, join));
                }
                return;
            }
            if let Some(entry) = graph.entry.as_ref() {
                converter
                    .dag
                    .add_edge(DAGEdge::state_machine_else(&branch_id, entry));
            }
            if let Some(join) = join_target {
                for exit_node in &graph.exits {
                    converter
                        .dag
                        .add_edge(DAGEdge::state_machine(exit_node, join));
                }
            }
        };

        connect_guarded_branch(self, if_guard, if_body_graph, join_id.as_ref());
        for (guard, graph) in elif_graphs {
            connect_guarded_branch(self, guard, graph, join_id.as_ref());
        }
        connect_else_branch(self, else_graph, join_id.as_ref());

        Ok(ConvertedSubgraph {
            entry: Some(branch_id),
            exits: join_id.into_iter().collect(),
            nodes,
            is_noop: false,
        })
    }

    /// Build a guard for elif branches: not prior_guards and current_condition.
    pub fn build_compound_guard(
        &self,
        prior_guards: &[ir::Expr],
        current_condition: Option<&ir::Expr>,
    ) -> Result<ir::Expr, DagConversionError> {
        let mut parts: Vec<ir::Expr> = Vec::new();
        for guard in prior_guards {
            let expr = ir::Expr {
                kind: Some(ir::expr::Kind::UnaryOp(Box::new(ir::UnaryOp {
                    op: ir::UnaryOperator::UnaryOpNot as i32,
                    operand: Some(Box::new(guard.clone())),
                }))),
                span: None,
            };
            parts.push(expr);
        }

        if let Some(condition) = current_condition {
            parts.push(condition.clone());
        }

        if parts.is_empty() {
            return Err(DagConversionError(
                "build_compound_guard called with no prior conditions and no current condition"
                    .to_string(),
            ));
        }
        if parts.len() == 1 {
            return Ok(parts.remove(0));
        }

        let mut result = parts.remove(0);
        for part in parts {
            result = ir::Expr {
                kind: Some(ir::expr::Kind::BinaryOp(Box::new(ir::BinaryOp {
                    left: Some(Box::new(result)),
                    op: ir::BinaryOperator::BinaryOpAnd as i32,
                    right: Some(Box::new(part)),
                }))),
                span: None,
            };
        }
        Ok(result)
    }
}

#[cfg(test)]
mod tests {
    use super::super::test_helpers::build_dag_with_pointers;

    #[test]
    fn test_convert_conditional_happy_path() {
        let dag = build_dag_with_pointers(
            r#"
            fn main(input: [x], output: [y]):
                if x > 0:
                    y = 1
                else:
                    y = 2
                return y
            "#,
        );

        assert!(
            dag.nodes.values().any(|node| node.node_type() == "branch"),
            "conditional should create a branch node"
        );
        assert!(
            dag.nodes.values().any(|node| node.node_type() == "join"),
            "conditional should create a join node"
        );
    }
}
